README.md
# Django Simplify

## Introduction

Django simplify provides Ruby on Rails-like command line functionalities, models and helper functions so you can focus on development and removes the pain of repeating frequent actions.

## Requirements

* Python `3.6` and above
* Django (tested with 3.1, probably works with any version that supports
  Python 3)

## Installation

`django-simplify` can be installed via pip.


```sh
$ pip install django-simplify
```

Then just add `simplify` to your `INSTALLED_APPS`.


Below are several example commands you can run.

```sh
$ python manage.py create_app <app_name>
```

## Features

* timestamp for every model
* alphabetic filter for admin
* model, view, template and url route generator
* automatic import


### Helper models

- simplify.helpers.model_helper.TimeBasedModel
  - Provides the `created_at` and `updated_at` fields for timestamp

- simplify.helpers.model_helper.NamedTimeBasedModel
  - Provides the `name`, `created_at` and `updated_at` fields.

- simplify.helpers.admin_helper.AlphaNumericFilterAdmin
  - when subclassed, it allows the items to be filtered alphabetically by either A-Z or 0-9
  - **Note**: for this to work, you must specify values for `alphanumeric_filter` in the model admin.
#### Usage

```python
# models.py
from simplify.helpers.model_helper import TimeBasedModel, NamedTimeBasedModel

class MyModel(TimeBasedModel):
    extra_fields = ....
```

```python
# admin.py
from simplify.helpers.admin_helper import AlphaNumericFilterAdmin

class MemberAdmin(AlphaNumericFilterAdmin):
    alphanumeric_filter = ["first_name", "last_name", 'age'] # this part is important
    list_filter = ['age']
    list_display = ['first_name', 'last_name',]
```


### Management commands

**1. create_app**

- automatically adds a `urls.py` file after app is created.
- adds newly created app in the settings.py file under `INSTALLED_APPS`
- creates `index`, `edit`, `create` and `detail` view and respective templates
- adds the app route to your project's `urls.py` file
  
#### Usage

```sh
$ python manage.py create_app <app_name>
```

**2. create_model**

* Creates a model and their respective fields

the following types maps to respective Django model fields

  - 121, o2o or set -> OneToOneField
  - bool -> BooleanField
  - date -> DateField
  - datetime or dt -> DateTimeField
  - dict or m2m -> ManyToManyField
  - email -> EmailField
  - file -> FileField
  - list or fk -> ForeignKey
  - float -> FloatField
  - dec -> DecimalField
  - img or image -> ImageField
  - int -> IntegerField
  - str or char -> CharField
  - txt or text -> TextField
  

#### Usage

```sh
$ python manage.py create_app <app_name> <model_name> field_name:type field_name:type ... 

# an example
$ python manage.py create_app member Member first_name:text last_name:text age:int
```

will generate the following code in the `member/models.py` file

```py
class Member(TimeBasedModel):
    first_name = models.CharField(max_length=50)
    last_name = models.CharField(max_length=50)
    age = models.IntegerField(default=0)

    def __str__(self):
        return self.first_name
```

> **Note**: The command uses the first specified field as the `__str__` default.


#### Specifying relationships

Specifying `ForeignKey`, `OneToOneField` or `ManyToManyField` is quite easy. just add an `=<related_model>`. See example

```sh
$ python manage.py create_app <app_name> <model_name> field_name:type=related_model


# an example
# if the related model is in the same models.py file, specify it as app_name.Model
$ python manage.py create_app author Author name:char books:fk=Book # or   
$ python manage.py create_app author Author name:char books:fk=author.Book 

# if in a different app. (say book model)
# obviously you should be able to substitute fk with m2m, o2o, 121
$ python manage.py create_app author Author name:char books:fk=book.Book 
```

will create the following

```py
class Author(TimeBasedModel):
    name = models.CharField(max_length=50)
    books = models.ForeignKey('book.Book', on_delete=models.CASCADE)

    def __str__(self):
        return self.name
```

## Todo
- add more helper functions
- add documentation



> **Note:** This is still in early development mode. might have bugs. It works fine if you write good code and follow the django style of development. Please fork the project to make contributions


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2] - 2021-05-06
### Removed
- removed model dependency on simplify so that you can still have your code running even when thesimplify package has been uninstalled and removed from `settings.INSTALLED_APPS`
- replaced `simplify.helpers.model_helper` and `simplify.helpers.admin_helper` with `simplify.utils`

### Added
- added `ListView` scaffold so you can include `l` in `crudil` parameter when running `python manage.py create_view crudil`
-
## [0.1.7 - 0.1.9] - 2021-04-14
### Fixes
- fixed simplify not importing timezone when `dt` or `date` is used as model field type
-
## [0.1.6] - 2021-04-14
### Added
- added app_name prefix to generated `urls` name in `app_name/urls.py`

## [0.1.5] - 2021-04-14
### Added
- added doc strings for `TimeBasedModel`, `TimeBasedModel` and `AlphaNumericFilterAdmin`

### Fixes
- fixed generated `urls.py` adding url paths on newline
- fixed `include` being imported multiple times for every app

## [0.1.4] - 2021-04-14
### Removed
- removed unnecessary user import statement in stub

### Fixes
- fixed `include` not being imported in `urls.py`
- fixed extra unnecessary new line in `urls.py` creation


## [0.1.3] - 2021-04-14
### Fixes
- fixed error caused due to extra format parameter added in 0.1.2
- replaced `User` model instead to use `_MY_MODEL_HERE_` placeholder

## [0.1.0 - 0.1.2] - 2021-04-13
### Added
- added create view functionality
- updated create_app so user can specify optional view to create
- added create_view command so user can create standalone view
- fixed error caused when inheriting from `AlphaNumericFilterAdmin` without specifying `alphanumeric_filter` fields.

## [0.0.6 - 0.0.9] - 2021-04-12
### Added
- added create model functionality
- added ability to auto generate admin model
- added `AlphaNumericFilterAdmin` to model helper

### Fixes
- bug fixes


## [0.0.2 - 0.0.5] - 2021-04-12
### Added
- added auto import of modules
- added ability to add created app to settings.INSTALLED_APPS
- added create_app command


## [0.0.1] - 2021-04-11
### Added
- Initial creation
- added `TimeBasedModel` and `NamedTimeBased` model


simplify/management/_helper.py
FIELD_MAPPERS = {
    '121': 'models.OneToOneField("{0}", on_delete=models.CASCADE)',
    'bool': 'models.BooleanField(default=True)',
    'char': 'models.CharField(max_length=50)',
    'date': 'models.DateField(default=timezone.now)',
    'datetime': 'models.DateTimeField(default=timezone.now)',
    'dt': 'models.DateTimeField(default=timezone.now)',
    'dict': 'models.ManyToManyField("{0}")',
    'email': 'models.EmailField(max_length=50)',
    'file': 'models.FileField()',
    'fk': 'models.ForeignKey("{0}", on_delete=models.CASCADE)',
    'float': 'models.FloatField(default=0)',
    'dec': 'models.DecimalField(default=0, max_digits=7, decimal_places=2)',
    'image': 'models.ImageField()',
    'img': 'models.ImageField()',
    'int': 'models.IntegerField(default=0)',
    'list': 'models.ForeignKey("{0}", on_delete=models.CASCADE)',
    'm2m': 'models.ManyToManyField("{0}")',
    'o2o': 'models.OneToOneField("{0}", on_delete=models.CASCADE)',
    'set': 'models.OneToOneField("{0}", on_delete=models.CASCADE)',
    'str': 'models.CharField(max_length=50)',
    'text': 'models.TextField()',
    'txt': 'models.TextField()',
}


MODEL_FIELD_TEMPLATE = """    {0} = {1}\n"""
